play-json/shared/src/main/scala/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -670,24 +671,37 @@ import scala.reflect.macros.blackbox
           val defaultValue = // not applicable for 'write' only
             defaultValueMap.get(name).filter(_ => methodName != "write")
 
+          val resolvedImpl = {
+            val implTpeName = Option(impl.tpe).fold("null")(_.toString)
+
+            if (implTpeName.startsWith(forwardPrefix) ||
+              (implTpeName.startsWith("play.api.libs.json") &&
```

Comment:
I don't think starting with `play.api.libs.json` proves it's a built-in format. If you use `Writes.apply` you'll have a class named something like `play.api.libs.json.Writes$$anon$6@509f0a9d`.

Maybe it's worth just skipping this check and doing the null check for everything?

Reply:
Are you sure? I'm pretty sure the lambda is defined in the class it's lexically inside. E.g. if you go to https://alvinalexander.com/scala/anonymous-classes-in-scala-examples and grep for AnonymousClassTest1 you'll see an example.

But I also have no problem with doing a null check for everything since the cost will be low.

Reply:
@cchantep can you verify this with a unit test?

Reply:
I may be mistaken here, but as I understand it `impl` is the `Tree` representing the implementation of the implicit `Writes`/`Reads`/`Format`. So if I have something like:

```scala
implicit val fooWrites = Writes[Foo] { foo => foo.value }
```

then Play is creating the anonymous class for you based on the lambda you passed.

Reply:
Such factory is instantiating from `FunctionalX` classes, not excluded from the null-check, contrary to other types from the package